### PR TITLE
Add MuJoCo environments to custom_mujoco/gym

### PR DESCRIPTION
**Pull request by anishhdiwan**
Added MuJoCo environments that implement the Gymnasium MuJoCo benchmark environments.

Added imitation learning algorithms GAIL, AIRL, AMP, NEAR, TRIRL

---
<a href="https://gitkraken.dev/review/github/nico-bohlinger/RL-X/pull/5?source=pr_review_chip">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://gitkraken.dev/images/figures/gitkraken-review-badge-dark.svg">
    Open with GitKraken
  </picture>
</a>

### CODE REVIEW

**Comment by nico-bohlinger** [comment]
Notes about cleanups that still have to be done:

1. There are some small syntax errors:
"config.on_demand_etas = Falsee"
"out_flat  = vmapped_chunk(inputs, params_chunk)c # (chunk_size, N, ...)"
also in algorithms due to missing imports:
"Tuple[float, float]"
Everything should compile properly. Please write uncommitted tests that actually runs everything for at least one step.

2. There are old imports that do not work in here:
"from trust_region_irl.algorithms.airl_ppo.flax_full_jit.general_properties import GeneralProperties"

3. There are numerous imports that are never used:
"from jax.tree_util import tree_map, tree_leaves"

4. There should be no type annotations:
"Tuple[float, float]"

5. Humanoid is missing forward after reset.

6. Do the mjx gym environments that we add really reproduce the official gym ones. Don't the official gym ones have frame skips? Can we make sure with uncomitted tests that everything is numerically identical (with some tiny tolerance for simulator differences). Also which one do we reproduce v4, v5?

7. All functions should always have two lines of space between each other. Same for imports and first function. This is currently not correct everywhere.

8. prepare_expert_data can return episode_returns and other optional keys, while expert_data_spec always declares rewards and a fixed key set. JAX io_callback requires these structures to match.

9. Please make the environment structure:
custom_mujoco/gym/humanoid_v5/mjx.

10. Make sure all READMEs are correctly linked in docs/README.md

11. Some paths contain your name:
"anishdiwan"

12. There are some files like "reward_correction.py" that are full of comments, unused imports, etc.

I see that the ## Resources section is missing in some algorithm readmes.

**Comment by nico-bohlinger** [comment]
@anishhdiwan I let my agent check your PR. Please fix the issues or let me know if some of the decisions here are deliberate. 

> Merge-blocking findings
> 1. [P1] GAIL, AMP, and AIRL train one discriminator but calculate policy rewards using the original discriminator state.
>    The local state is updated, but the reward calls pass self.discriminator_state:
>    - [GAIL lines 288–316](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/gail_ppo/flax_full_jit/gail_ppo.py#L288-L316)
>    - [AMP lines 288–316](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/amp_ppo/flax_full_jit/amp_ppo.py#L288-L316)
>    - [AIRL lines 305–336](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/airl_ppo/flax_full_jit/airl_ppo.py#L305-L336)
>    Consequently, PPO learns against the initial random reward model. The direct fix is replacing those six uses with the local discriminator_state, followed by a regression test proving that rewards change after a discriminator update.
> 2. [P1] Three algorithms are not runnable.
>    - TRIRL-DTRL imports two nonexistent modules at [lines 21–22](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/trirl_dtrl/flax_full_jit/trirl_dtrl.py#L21-L22). Both should reference flax_full_jit.
>    - AIRL reads an absent retraining config value, never initializes nr_epochs_disc, and initializes log_prob with a shape that breaks the discriminator for multidimensional actions: [constructor lines 55–139](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/airl_ppo/flax_full_jit/airl_ppo.py#L55-L139).
>    - NEAR uses uninitialized H_terminal and handle_absorbing_states on its first training trace: [lines 393–415](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/near_ppo/flax_full_jit/near_ppo.py#L393-L415).
> 3. [P1] The advertised defaults cannot complete an update.
>    Ant defaults produce 4000 × 10 = 40,000 rollout samples, while minibatch_size=512 floors the number of minibatches to 78. The code then attempts to reshape 400,000 indices into 399,360 elements. The supplied Ant expert dataset has only 30,000 transitions as well, but it is indexed as though it had 40,000.
>    This duplicated logic affects GAIL, AMP, AIRL, and both TRIRLs. Expert examples should be sampled with replacement, and batch divisibility must be validated or the remainder explicitly dropped.
> 4. [P1] The locomotion environments still do not reproduce Gymnasium v5.
>    Frame skips and reset-time mjx.forward are fixed. However:
>    - All five locomotion environments use final qvel[0] for forward reward. Gymnasium v5 uses position or body-COM displacement over the complete frame-skipped step.
>    - Ant, Hopper, Humanoid, and Walker2d retain the old healthy-reward behavior, awarding survival reward on unhealthy terminal transitions. V5 uses healthy_reward * is_healthy.
>    - Ant omits the full-state finiteness check; Hopper omits healthy_state_range=(-100, 100).
>    With aligned MuJoCo/MJX states, next observations agreed within roughly 5e-8–2e-5, while one-step rewards differed by 0.0053–0.0626, isolating this as formula error rather than simulator drift. The official implementation is visible in [Gymnasium’s Hopper-v5 source](https://github.com/Farama-Foundation/Gymnasium/blob/main/gymnasium/envs/mujoco/hopper_v5.py).
> 5. [P1 integration] The environments are already incompatible with current RL-X algorithms after rebasing.
>    Every new [BoxSpace](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/environments/custom_mujoco/gym/ant_v5/mjx/box_space.py#L4-L12) lacks the center and scale fields now used by full-JIT FastSAC and FastMPO. I reproduced an AttributeError using PR Ant with current FastSAC. These copies should be synchronized with the current shared behavior.
> 6. [P1 semantic] The current AMP implementation should not be presented or ported as AMP.
>    Its discriminator is D(state, action) and explicitly ignores next_state: [discriminator lines 21–39](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/amp_ppo/flax_full_jit/discriminator.py#L21-L39). It is effectively least-squares GAIL. AMP is based on motion/state-transition features and combines its style reward with a task objective, as described in the [AMP paper](https://arxiv.org/abs/2104.02180). Here env_reward_frac is also unused.
> Other correctness issues include:
> - env_reward_frac is inert in all six algorithms.
> - TRIRL-DTRL passes the opposite maximum_eta setting at [lines 353–361](https://github.com/nico-bohlinger/RL-X/blob/e270942a42c1dd14bdceb8100088e5354b8a242c/rl_x/algorithms/trirl_dtrl/flax_full_jit/trirl_dtrl.py#L353-L361).
> - Both TRIRLs multiply entropy regularization by the coefficient twice.
> - GAIL, AMP, and both TRIRLs omit learned reward/discriminator and buffer state from checkpoints, so training cannot resume faithfully.
> - prepare_expert_data can flatten trajectories to a size different from the callback specification and does not reliably match its declared float32 shapes.
> - PointMaze is a custom task, not the advertised Gymnasium-Robotics PointMaze; observation, reset, reward, success radius, and termination semantics differ substantially.

**Comment by anishhdiwan** [comment]
Hey, I made two new commits to address your comments. Also let an agent make some general fixes and check the code. Here's a summary:

**1. Discriminator state for rewards**: Rewards were computed from the freshly initialised discriminator rather than the one being trained. The reward now uses the trained discriminator. Bug was introduced in manual porting of paper code.

**2. Three algorithms not runnable**: TRIRL-DTRL imported modules that do not exist, AIRL read a config key that was never defined and built its discriminator input with the wrong shape, and NEAR used two attributes it never created. All are now defined and all six algorithms run.

**3. Defaults cannot complete an update**: The reshape into minibatches failed whenever the minibatch size did not divide the batch evenly, and the expert set was indexed as though it were as large as the rollout. The leftover samples are now dropped before reshaping, and expert samples are drawn randomly with replacement, so the size of the expert dataset no longer matters.

**4. Locomotion environments do not reproduce v5**: The forward reward used the instantaneous velocity at the end of the step instead of the displacement across the whole step, survival reward was still paid on the terminal unhealthy step, and Ant and Hopper were each missing a health check. All three now follow the v5 formulas.

**5. BoxSpace missing `center` and `scale`**: The six copied action spaces predated fields that FastSAC and FastMPO now read, so those algorithms crashed on these environments. Every copy has been replaced with the current shared version.

**6. AMP**: This is a design choice. The original paper uses state-next_state rewards and a fixed fraction of env reward. Both are possible with the PR code by changing options in the config. However, the defaults are set to state-action reward with no env reward for fair comparison with other methods. 

**7. `env_reward_frac` inert in all six**: Same reason as above. Env rewards can be used but aren't enabled by default since they skew the IRL/IL algorithm's true performance.

**8. TRIRL-DTRL opposite `maximum_eta`**: The flag was flipped during a manual code rewrite. The negation was removed.

**9. Both TRIRLs apply the entropy coefficient twice**: Done on purpose. The factor is also added to the learnt reward and mainly acts to reduce the critic MSE loss. It does not change the optimization problem at all since its just a cost * (reward + entropy) optimization problem. 

**10. Checkpoints omit discriminator and buffer state**: Only the policy and critic were saved, so a resumed run lost the learned reward. The discriminator is now saved too, and TRIRL additionally saves either its reward network or its buffers, depending on which one it is using.

**11. `prepare_expert_data` size and type mismatch**: The expert data was not cast to the type the loader declares, and the sample count was read from the file before the trajectories were flattened, so the two could disagree. Both now come from the flattened, cast arrays.

**12. PointMaze is not the Gymnasium-Robotics PointMaze**:  The README now describes it as a custom task and not a port.

**Comment by nico-bohlinger** [comment]
Thanks for the updates. There are still a few issues to address:

1. **AMP and `env_reward_frac` (points 6–7):** I can’t find the configuration options you mention in the submitted code. AMP’s discriminator still hard-codes state–action inputs and ignores the next state. Also, `env_reward_frac` is only assigned in the constructors and never used in any of the six algorithms, so changing it currently has no effect. Keeping environment reward disabled by default is fine, but the option needs to work. AMP also uses `0` for policy samples, whereas the paper’s least-squares objective uses `-1`, with `+1` for expert samples.

2. **TRIRL entropy scaling (point 9):** I understand the intended common scaling, but the current implementation doesn’t seem to apply it consistently. `beta = 1 / entropy_coef`, and the reward correction multiplies by both `beta` and `entropy_coef`, so those factors cancel. Meanwhile, entropy is multiplied by `entropy_coef` twice. That changes the relative weighting rather than just scaling the whole objective. Could you check this against the intended objective?

3. **Checkpoints (point 10):** When `reward_fn_approximator=True`, the discriminator/eta buffers are still omitted, although training continues to use them to generate the reward-network targets. DTRL also needs its policy history when `on_demand_etas=True`. These should be restored when resuming training.

4. **Expert data (point 11):** The dtype and flattened sample count fixes work, but `episode_returns` still has one entry per episode while the callback expects one per transition. This still produces a callback shape error for trajectory-shaped datasets. Since the algorithms don’t use expert rewards, removing rewards from the callback and its specification seems simplest.

5. **Environment defaults:** The new environments should also work with the existing RL-X algorithms. With current full-JIT PPO defaults, `nr_envs=4000` fails the evaluation-frequency divisibility check and doesn’t divide cleanly into minibatches. PointMaze’s `64` environments produce zero PPO minibatches. Using `4096` consistently would be a straightforward fix.

6. **Two smaller correctness fixes:** AIRL passes `1 / (high-low)` where an absorbing-state **log probability** is expected. Also, discriminator learning-rate annealing uses the wrong denominator when policy and discriminator epoch counts differ.

Finally, RL-X now has both **PyTorch and Flax implementations**, so I would also expect both frameworks for the algorithms we add in this PR. Currently these only have `flax_full_jit`; please also add the regular `flax` and `pytorch` implementations.

Please resolve the README conflicts with current `master` as well.

**Comment by anishhdiwan** [comment]
Hey, thanks. Made the suggested changes. Regarding points 2,3:
- The entropy scaling is consistent with the objective and optimal reward equations (eq1, eq2) in https://arxiv.org/pdf/2605.11020. The optimal reward is `beta*disc` which blows up the critic MSE loss. The fix was to scale the whole problem by entropy coefficient so that the high reward scale is reduced (this means that double entropy coefficient scaling of the entropy term is not a bug).
- Setting `reward_fn_approximator=True` trains a separate network to regress the interpolated discriminator buffer. This does mean that restarting training from a checkpoint starts off with empty buffers, but the buffer is also intitalized with the output of the reward approximator, so it should not have a big impact on training.
